### ISSUE TITLE
Make modular code for an experiment

### ISSUE THREAD

**Issue by nicholas-denis** [nicholas-denis/ppi-testing#10]
The code in the notebook is perfectly fine for testing and prototyping, but it doesn't scale. It will be a great asset to learn and practice, in all your coding projects (trust me, there is overhead in developing the skill, but it will pay off incredibly) to write small modular code from the start.

Below (at bottom) is an example of an experiment from your notebook. I will comment and address how to make it more modular.

- [x] with config file, many of the "constants" or values are in the config, so those lines disappear
- [x] second_pop_params = some_helper_function_that_gets_second_pop_params(config)
- [x] ** SIDE NOTE **, rather than getting a value from a dictionary key using config[my_key], use config.get('my_key', <HERE>). It takes a bit more to type and has the annoying effect of stretching your fingers on the keyboard to get the ' key (annoying for me), BUT, it is safer.  Whatever you put in <HERE> will be the default value if 'my_key' is not actually in config. if you just do config.get('my_key'), then it defaults to None.
- [x] We can figure out what the some_helper_function_that_gets_second_pop_params(args) looks like, but we are not hard coding it
- [x] often times in experiments, we start with a couple of lists to capture metrics.. but over time as we do more and more experiments we start to collect more and more metrics... next thing you know "ppi_mean_widths = []" becomes 10 different lists.... let's have a function "def get_init_metrics_dict(config): ", and it returns a dictionary that has:
my_dict.get('ppi_mean_widths') # dict, and all the other lists you are tracking. That way at the end of the experiment, when we return all these lists or do plots, we can just pass the single dict. Moreover, when we decide later to add another metric, we just go to that function, add it to the dict, etc... it is a bit cleaner
- [x] in a general sense, there is a tradeoff with verbosity and distinct semantics of a variables. We don't want something like the_mean_ci_width_for_PPI_plus_plus... but we want to make sure that we can easily distinguish the meaning of two variables, like "ppi_mean_widths " and "ppi_widths". If there is hierarchy, for example some are for 'ppi_experiment', some are for 'naive_experiment', and some are for 'classical' and the ppi_mean_widths is for... ??(I can't tell from the variable names), then in our metrics_dict, we can have something like metrics_dict = {'experimental_settings': 
                                                                                                                              {'ppi': 
                                                                                                                                      {'ci_widths' : [ ],
                                                                                                                                        ... },
                                                                                                                               {'naive': 
                                                                                                                                      {'ci_widths' : [ ],
                                                                                                                                        ... },
...
This will have a clear semantics, and though it can get quite hierarchical, I think it is easier to track and harder to get mixed up.

- [  ] I won't go over all of the modular stuff, hopefully you can get the drift. You can always share your code to gpt4o or claude opus, and ask it to recommend refactoring, making it more modular, use best practices, etc, BUT I will say, the blocks below where you do a bunch of similar operations for ppi/classical/naive, those should all be function calls
- [  ] NOTE ON USE OF student's t-distribution vs Normal: Let's make sure this is covered/represented as a boolean in the config yaml file. We will want to set it in config, and so the output should just check the boolean flag and use the appropriate approach.

ppi_mean_widths = []
naive_mean_widths = []
classical_mean_widths = []

# Iterations

n_its = 10

for params in second_pop_params:
    alpha_1, beta_2 = params

    ppi_preds = []
    ppi_lowers = []
    ppi_uppers = []
    ppi_true_bias = []
    ppi_covered = []
    ppi_widths = []

    naive_preds = []
    naive_lowers = []
    naive_uppers = []
    naive_covered = []
    naive_widths = []

    classical_preds = []
    classical_lowers = []
    classical_uppers = []
    classical_covered = []
    classical_widths = []

    residuals = []
    biases = []

    true_mean = beta_2**2 * alpha_2 * (alpha_2 + 1)
    for i in range(n_its):
        # Sampling

        x_train, y_train = sample_mv_norm_gamma_population(alpha_1, beta_1, train_sample, vec) 

        x_gold, y_gold = sample_mv_norm_gamma_population(alpha_2, beta_2, small_sample, vec)

        x_ppi, y_ppi = sample_mv_norm_gamma_population(alpha_2, beta_2, large_sample, vec)

        x_train, x_test, y_train, y_test = train_test_split(x_train, y_train, test_size=0.2)

        y_train, y_test = y_train.flatten(), y_test.flatten()

    
        start = time.time()
        
        model = build_fit_rf(x_train, y_train)  # Training

        end = time.time()

        print(end - start)

        # Residual testing

        y_test_pred = model.predict(x_test)
        residual = np.sqrt(np.mean((y_test_pred - y_test)**2))  # Mean squared deviation
        residuals.append(residual)

        y_gold_fitted = model.predict(x_gold)  # Gold standard fitted
        y_fitted = model.predict(x_ppi)  # Unlabelled fitted

        # Manual rectifier computation
        rectifier = np.mean(y_gold_fitted - y_gold)

        # PPI section 

        ppi_theta = ppi_py.ppi_mean_pointestimate(y_gold, y_gold_fitted, y_fitted, lhat=1)
        ppi_theta_ci = ppi_py.ppi_mean_ci(y_gold, y_gold_fitted, y_fitted, lhat=1)

        ppi_preds.append(ppi_theta[0])
        ppi_lowers.append(ppi_theta_ci[0][0])
        ppi_uppers.append(ppi_theta_ci[1][0])
        ppi_true_bias.append(np.mean(y_fitted - y_ppi))
        if true_mean >= ppi_theta_ci[0][0] and true_mean <= ppi_theta_ci[1][0]:
            ppi_covered.append(1)
        else:
            ppi_covered.append(0)
        ppi_widths.append(ppi_theta_ci[1][0] - ppi_theta_ci[0][0])
        biases.append(np.mean(y_fitted - y_ppi))

        # Naive imputation
        concat = np.vstack((y_gold, y_fitted.reshape(-1,1))).flatten()  # Concactenate samples
        naive_theta = np.mean(concat)  # Compute mean
        naive_sigma = np.std(concat)  # Compute std dev
        n_tot = concat.shape[0]
        naive_theta_ci = stats.norm.interval(0.9, loc=naive_theta, scale=naive_sigma/np.sqrt(n_tot))  # Use norm as N is large

        naive_preds.append(naive_theta)
        naive_lowers.append(naive_theta_ci[0])
        naive_uppers.append(naive_theta_ci[1])
        if true_mean >= naive_theta_ci[0] and true_mean <= naive_theta_ci[1]:
            naive_covered.append(1)
        else:
            naive_covered.append(0)
        naive_widths.append(naive_theta_ci[1] - naive_theta_ci[0])

        # Classical prediction (Gold standard only)
        
        classical_theta, classical_se = np.mean(y_gold.flatten()), stats.sem(y_gold.flatten())
        h = classical_se * stats.t.ppf((1 + .9) / 2., small_sample-1)  # Highly stolen code, uses t-dist here

        classical_preds.append(classical_theta)
        classical_lowers.append(classical_theta - h)
        classical_uppers.append(classical_theta + h)
        if true_mean >= classical_theta - h and true_mean <= classical_theta + h:
            classical_covered.append(1)
        else:
            classical_covered.append(0)
        classical_widths.append(2 * h)

    print("Rho: ", rho)
    print("Average residual: ", np.mean(residuals), "True bias :", np.mean(biases))
    print("PPI.", "Average pred: ", np.mean(ppi_preds), "Average lower bound: ", np.mean(
        ppi_lowers), "Average upper bound: ", np.mean(ppi_uppers), "Average width: ", np.mean(ppi_widths), "Covered percent: ", np.mean(ppi_covered))
    print("Naive.", "Average pred: ", np.mean(naive_preds), "Average lower bound: ", np.mean(
        naive_lowers), "Average upper bound: ", np.mean(naive_uppers), "Average width: ", np.mean(naive_widths), "Covered percent: ", np.mean(naive_covered))
    print("Classical.",  "Average pred: ", np.mean(classical_preds), "Average lower bound: ", np.mean(
        classical_lowers), "Average upper bound: ", np.mean(classical_uppers), "Average width: ", np.mean(classical_widths), "Covered percent: ", np.mean(classical_covered))

    ppi_mean_widths.append(np.mean(ppi_widths))
    naive_mean_widths.append(np.mean(naive_widths))
    classical_mean_widths.append(np.mean(classical_mean_widths))
